sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -64,10 +64,17 @@ object PhysicalOperation extends PredicateHelper {
         val substitutedFields = fields.map(substitute(aliases)).asInstanceOf[Seq[NamedExpression]]
         (Some(substitutedFields), filters, other, collectAliases(substitutedFields))
 
-      case Filter(condition, child) if condition.deterministic =>
+      case Filter(condition, child) =>
         val (fields, filters, other, aliases) = collectProjectsAndFilters(child)
-        val substitutedCondition = substitute(aliases)(condition)
-        (fields, filters ++ splitConjunctivePredicates(substitutedCondition), other, aliases)
+
+        // Deterministic parts of filter condition placed before non-deterministic predicates could
+        // be pushed down safely.
+        val (pushDown, rest) =
```

Comment:
I also think that silently dropping nondeterministic filters can be dangerous. Maybe we should just return all operators beneath the top-most nondeterministic filter as the bottom operator?

For example, say we have a plan tree like this:

```
Project a, b
 Filter a > 1
  Filter b < 3
   Filter RAND(42) > 0.5
    Filter c < 2
     TableScan t
```

We should return the following result:

```
(
  // Project list
  Seq(a, b),

  // Deterministic filters
  Seq(b < 3, a > 1),

  // The top-most nondeterministic operator filter with all operators beneath
  Filter(
    RAND(42) > 0.5,
    Filter(c < 2,
      TableScan(t)))
)
```